src/sync/deduplicator.ts

```diff
@@ -0,0 +1,200 @@
+import {Redis} from "ioredis";
+
+/**
+ *
+ * The generic idea of the Deduplicator(tm) is very simple: while executing the job, Deduplicator raises a flag (in redis)
+ * to inform other executors not to run the same job ("same"-ness is defined by a key). At the end of the execution it
+ * removes the previously raised flag.
+ *
+ * As all "very simple" ideas, it has several challenges, namely:
+ *
+ * THE FIRST CHALLENGE is: what if the executor dies (e.g. the EC2 node is rotated) during the execution? This might
+ * have a flag raised but not being deleted.
+ *
+ * The implemented solution: save the timestamp of the flag and keep updating it in the executor while it is doing its
+ * job. If it dies, the flag will be invalidated in some reasonable time.
```

Comment:
I guess using normal TTL would do the job, hence in the unlike situation of EC2 death, we'll only delay execution for a while (Then the message will be retried and TTL expired)

If we'll be updating time as we process we close the door for having retries on the timeout hence if we recently updated the timestamp on retry the message can be discarded

Reply:
The problem with a single TTL is that it is unclear how long the job takes to execute. We might drop TTL before the job is finished, which will bring us back to duplication problem.

Having a preriodically updating flag looked more robust to me.